// furry-wrapper/dist/Base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = require("axios");
class Base {
    constructor(options = {}) {
        this.options = options;
        this.token = options.token || options.auth;
    }
    static get defaultAgent() {
        return `Node ${process.version} @ ${process.platform}`;
    }
    get agent() {
        return this.options.agent || Base.defaultAgent;
    }
    query(url, cb) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            if (!cb)
                cb = r => r.data;
            const ops = ((_a = this.options) === null || _a === void 0 ? void 0 : _a.axiosOptions) || {};
            if (!ops.headers)
                ops.headers = {};
            ops.headers['User-Agent'] = ((_b = this.options) === null || _b === void 0 ? void 0 : _b.agent) || Base.defaultAgent;
            if (!ops.timeout)
                ops.timeout = 10000;
            if (this.token)
                ops.headers['Authorization'] = `Bearer ${this.token}`;
            return yield axios_1.default.get(url, ops)
                .then(cb)
                .catch(e => e);
        });
    }
    /* eslint-disable @typescript-eslint/ban-types */
    static bundle(base, items, options) {
        const bundler = {};
        Object.keys(items).forEach((name) => {
            bundler[name] = { value: (options === null || options === void 0 ? void 0 : options.default) ? options.default(items[name]) : (function (options) {
                    return __awaiter(this, void 0, void 0, function* () {
                        return yield base(items[name], options);
                    });
                }) };
        });
        return Object.defineProperties(Object.values(bundler)[0].value, bundler);
    }
}
exports.default = Base;
